growattServer/__init__.py

```diff
@@ -144,17 +144,219 @@ def tlx_detail(self, tlx_id):
+                                      'pdischarge': '0',
+                                      'ppv': '0',
+                                      'sysOut': '0.93'},
+                          ......
+                     }
+        'eAcCharge' -- Exported to grid in kWh
+        'eCharge' -- System production in kWh = Self-consumption + Exported to Grid
+        'eChargeToday' -- Load consumption from solar in kWh
+        'eChargeToday1' -- Self-consumption in kWh
+        'eChargeToday2' -- Self-consumption in kWh (eChargeToday + echarge1)
+        'echarge1' -- Load consumption from battery in kWh
+        'echargeToat' -- Total battery discharged (all time) in kWh
+        'elocalLoad' -- Load consumption in kW (battery + solar + imported)
+        'etouser' -- Load consumption imported from grid in kWh
+        'photovoltaic' -- Load consumption from solar in kWh (same as eChargeToday)
+        'ratio1' -- % of system production that is self-consumed
+        'ratio2' -- % of system production that is exported
+        'ratio3' -- % of Load consumption that is "self consumption"
+        'ratio4' -- % of Load consumption that is "imported from grid"
+        'ratio5' -- % of Self consumption that is directly from Solar
+        'ratio6' -- % of Self consumption that is from batteries
+        'unit' -- Unit of measurement e.g kWh
+        'unit2' -- Unit of measurement e.g kW
+
+
+        NOTE - It is possible to calculate the PV generation that went into charging the batteries by performing the following calculation:
+        Solar to Battery = Solar Generation - Export to Grid - Load consumption from solar
+                           epvToday (from mix_info) - eAcCharge - eChargeToday
+        """
+        assert timespan in Timespan
+        if timespan == Timespan.day or timespan == Timespan.hour:
+            date_str = date.strftime('%Y-%m-%d')
+        elif timespan == Timespan.month:
+            date_str = date.strftime('%Y-%m')
+
+        response = self.session.post(self.get_url('newMixApi.do'), params={
+            'op': 'getEnergyProdAndCons_KW',
+            'plantId': plant_id,
+            'mixId': mix_id,
+            'type': timespan.value,
+            'date': date_str
+        })
+        data = json.loads(response.content.decode('utf-8'))
+
+        #Some of the 'totals' values that are returned by this function do not align to what we would expect, however the graph data always seems to be accurate.
+        #Therefore, here we take a moment to calculate the same values provided elsewhere but based on the graph data instead
+        #The particular stats that we question are 'load consumption' (elocalLoad)  and 'import from grid' (etouser) which seem to be calculated from one-another
+        #It would appear that 'etouser' is calculated on the backend incorrectly for systems that use AC battery charged (e.g. during cheap nighttime rates)
+        if timespan == Timespan.hour or timespan == Timespan.day:
```

Comment:
I'm happy to remove it and put it into the HomeAssistant library and my example file only. 

HOWEVER, note that I do not modify any of the values that are returned, I just add new ones to the data structure that is returned from the API. I thought it would be helpful for people to have this done for them, they can then choose whether or not they want to use it.

What would you like me to do? I'm happy either way, I'd just like to get it merged in :)

Reply:
@indykoning any thoughts on how you'd like to proceed on this one?

Reply:
Id like to keep any custom code at a minimum unless it's for backwards compatible reasons

Reply:
No problem at all. I completely understand. 

I'll remove these additions to the function and make it just return the raw values with no additions. I'll move this logic into my example file so that others can see how to calculate these values if they wish.